### PR TITLE
imx6ull: watchdog handling

### PR DESCRIPTION
**Pull request by jsarzy**
## Description

Add `wdgreload` syscall for imx6ull platform. Enable a watchdog timer in plugin.

PR requires changes in the psu scripts, so I will merge it by myself.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: armv7a-imx6ull

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by agkaminski — APPROVED** [review]
Have you tested that if WATCHDOG is not defined watchdog remains not enabled?

**Comment by jsarzy** [comment]
> Have you tested that if WATCHDOG is not defined watchdog remains not enabled?

Yep, it remains not enabled.

**Review by nalajcie — APPROVED** [review]

File: hal/armv7a/imx6ull/_init.S [hunk 1]
```diff
@@ -88,6 +88,17 @@ plugin:
 	/* note: "ldr reg, =constant" pseudo instruction can't be used before
 	 * the serial download check below - the literal pool will not be included in the plugin */
 
+	/* init watchdog */
+	mov32 r4, 0x20bc000
+#if defined(WATCHDOG)
+	/* 128 sec timeout, the clock is already enabled */
+	movw r5, #0xff34
```
Comment (nalajcie) [inline hal/armv7a/imx6ull/_init.S:95]: can we change it to `#0xff36` to enable WDBG bit by default (_Watchdog DEBUG Enable. Determines the operation of the WDOG during DEBUG mode_ - 1 means Suspend WDG during DEBUG mode).

Makes debugging using JLink much easier :)

File: hal/armv7a/imx6ull/_init.S [hunk 2]
```diff
@@ -88,6 +88,17 @@ plugin:
...
+#else
+	/* if watchdog is disabled feed him default value */
```
Comment (nalajcie) [inline hal/armv7a/imx6ull/_init.S:97]: Maybe add comment that it's needed to keep `plugin` section size the same regardless of `WATCHDOG` define?

**Review by nalajcie — CHANGES_REQUESTED** [review]

File: hal/armv7a/imx6ull/_init.S (hunk 1 above)
Comment (agkaminski) [reply hal/armv7a/imx6ull/_init.S:95]: Does it have any side effects? If so, we might want to do this only if `NDEBUG` is not defined

File: hal/armv7a/imx6ull/_init.S (hunk 2 above)
Comment (agkaminski) [reply hal/armv7a/imx6ull/_init.S:97]: Good idea, it might get removed if no rationale is presented

File: hal/armv7a/imx6ull/_init.S (hunk 1 above)
Comment (nalajcie) [reply hal/armv7a/imx6ull/_init.S:95]: no side effects (affects only DEBUG mode). Still I would like to have it always enabled I might want to debug userspace app regardless of kernel build version.

Alternative approach is to disable wdog in gdb (only allowed in DEBUG mode):
```gdb
set *imx6ull_common.wdog &= ~0x4
```

**Comment by nalajcie** [comment]
There is also a side issue - upon enabling watchdog on `imx6ull` our current `reboot` implementation is no longer valid.

I don't want to create separate PR as this is closely related, maybe You can add the below change:
```diff
--- a/hal/armv7a/imx6ull/imx6ull.c
+++ b/hal/armv7a/imx6ull/imx6ull.c
@@ -316,7 +316,8 @@ static int _imx6ull_getIOisel(int isel, char *daisy)

 static void _imx6ull_reboot(void)
 {
-       *(imx6ull_common.wdog + wdog_wcr) |= 4;
+       /* assert SRS signal by writing 0 to bit 4 and 1 to bit 2 (WDOG enable) */
+       *(imx6ull_common.wdog + wdog_wcr) = (1 << 2);
        for (;;) ;
 }
```

**Comment by jsarzy** [comment]
> There is also a side issue - upon enabling watchdog on `imx6ull` our current `reboot` implementation is no longer valid.
> 
> I don't want to create separate PR as this is closely related, maybe You can add the below change:
> 
> ```diff
> --- a/hal/armv7a/imx6ull/imx6ull.c
> +++ b/hal/armv7a/imx6ull/imx6ull.c
> @@ -316,7 +316,8 @@ static int _imx6ull_getIOisel(int isel, char *daisy)
> 
>  static void _imx6ull_reboot(void)
>  {
> -       *(imx6ull_common.wdog + wdog_wcr) |= 4;
> +       /* assert SRS signal by writing 0 to bit 4 and 1 to bit 2 (WDOG enable) */
> +       *(imx6ull_common.wdog + wdog_wcr) = (1 << 2);
>         for (;;) ;
>  }
> ```

Sure, I'll add.

**Review by nalajcie — APPROVED** [review]

**Pull request merged**